### PR TITLE
[graph] Add graph core class 

### PR DESCRIPTION
**Pull request by kparichay**
Add graph core class
This is mainly separating the existing implementation of core part
of the graph from NetworkGraph.
This patch does not contain new implementation but mostly contain
borrowed implementation from existing graph.

See Also #1176

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/layers/layer_node.h [hunk 1]
```diff
@@ -69,12 +69,21 @@ class LayerNode : public ml::train::Layer, public GraphNode {
...
   }
 
+  /**
+   * @brief     set name of layer
+   *
+   * @param[in] name Name of the layer
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+   */
+  int setName(const std::string &name) { return setProperty({"name=", name}); }
```
Comment (zhoonit) [inline nntrainer/layers/layer_node.h:85]: maybe this?
```suggestion
  int setName(const std::string &name) { return setProperty({"name=" + name}); }
```

File: nntrainer/graph/graph_core.h [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    typename std::vector<std::shared_ptr<GraphNode>>::const_reverse_iterator
+      const_reverse_iterator;
+
+  /**
+   * @brief     Constructor of Graph Core Class
+   */
+  GraphCore() : def_name_count(0) {}
+
+  /**
+   * @brief Create new GraphNode and add into Graph
```
Comment (zhoonit) [inline nntrainer/graph/graph_core.h:53]: is this also creating a new graphnode?

File: nntrainer/layers/layer_node.h (hunk 1 above)
Comment (kparichay) [reply nntrainer/layers/layer_node.h:85]: yes, this was done in the next PR.

Comment (kparichay) [reply nntrainer/layers/layer_node.h]: done here as well.

File: nntrainer/graph/graph_core.h (hunk 2 above)
Comment (kparichay) [reply nntrainer/graph/graph_core.h]: no, it is not. Updated documentation.

**Review by zhoonit — APPROVED** [review]
LGTM

**Pull request merged**